why-did-engine-decide-not-to-use-the-zone-map.md

PeggySchmidtMittenzwei opened the pull request:
https://exasol.atlassian.net/browse/L3-3110

PeggySchmidtMittenzwei committed:
Create zone-map-usage-optimizer-question.md

PeggySchmidtMittenzwei committed:
Update zone-map-usage-optimizer-question.md

PeggySchmidtMittenzwei committed:
Update zone-map-usage-optimizer-question.md

fixed 
Multiple consecutive blank lines
Trailing spaces

PeggySchmidtMittenzwei committed:
Update zone-map-usage-optimizer-question.md

fixed
Trailing spaces
Blank line inside blockquote

exaSR commented:
1. as usual, the main heading in the article should/must match the article file name
2. I would recommend to put the "summary" part very early and put the whole complex explanation below that. Also, maybe the "Authors" towards the end of the article.

PeggySchmidtMittenzwei committed:
Update zone-map-usage-optimizer-question.md

incoopertaed @exaSR comments

as usual, the main heading in the article should/must match the article file name
I would recommend to put the "summary" part very early and put the whole complex explanation below that. Also, maybe the "Authors" towards the end of the article.

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,198 @@
+# Why did the optimizer decide not to use the zone map?
```
In this case, for this example, it is Engine that chooses whether to use zone maps or not. You already mentioned that in slack :grin:

narmion replied:
Already outdated.

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,198 @@
+
+## Question
+
+Even if you force zone map usage and our WHERE clause is prunable, the optimizer might still avoid using the zone map:
```
You never force zone map usage. You force the existence of zone maps and their maintenance. 
And as above, the Engine decides not the optimizer ;).

narmion replied:
Already outdated.

PeggySchmidtMittenzwei committed:
Update zone-map-usage-optimizer-question.md

replace optimizer with engine

PeggySchmidtMittenzwei committed:
Update zone-map-usage-optimizer-question.md

added:
The engine computes the data based on the execution plan created by the compiler. During filter stages, the Engine decides on the use of zone maps.

PeggySchmidtMittenzwei committed:
Update zone-map-usage-optimizer-question.md

I adapted the data generation from RANDOM to round(RANGE_VALUE mod 4*365,0)

with this with this it is more clearly visible

PeggySchmidtMittenzwei committed:
Update zone-map-usage-optimizer-question.md

PeggySchmidtMittenzwei committed:
Update zone-map-usage-optimizer-question.md

PeggySchmidtMittenzwei committed:
Update zone-map-usage-optimizer-question.md

replaced filter to scan (public profile)

PeggySchmidtMittenzwei committed:
Update zone-map-usage-optimizer-question.md

corrected links
added profiling links

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+
+Let us go the first execution:
+
+The SCAN uses the index on SID to identify the best ROWID order for the GROUP BY. So the data block layout after
```
Perhaps:
Let us assume the data block layout after the SCAN is:

The real ids from the example would be hard to get and hard to make it so obvious. Note that segment ids are not visible directly anywhere.

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+
+## Question
+
+Even if you force zone map usage and our WHERE clause is prunable, engine might still avoid using the zone map:
```
"you force zone map usage"

You can only force zone map existence on a column. You cannot enforce usage.

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+# Why did engine decide not to use the zone map?
+
+* even though zone map is forced and
```
"even though a zone map exists" is better as the "forced" might be misinterpreted.

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
````diff
@@ -0,0 +1,201 @@
+    VALUES BETWEEN 1 AND 1000000*200*NPROC()+1;
+```
+
+We also include the following data to better explain the engine’s decisions:
````
I would not add an INSERT of this data here. This is an artificial example from the very beginning.

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+
+### Explanation of the execution plan without zone map usage
+
+Let us go the first execution:
```
Let us go the first execution:
->
Let us examine the first execution with an artificial data layout:

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+Accessing the metadata of a segment has an overhead and also consumes memory.  So our engine only accesses the metadata if consecutive rows in the data block have the same SEGMENT_ID.
+
+> [!NOTE]
+> Due to that mixed up TIMES_SEGMENT_IDs, our database engine does not use zone maps for the SCAN although zone maps are available.
```
available.

Add the following:
This is necessary, because the use of zone maps would increase the execution time instead of reducing it.

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+|65|6|'2025-06-05'|$${\color{red}5}$$|
+
+Now we do a GROUP BY POSTPROCESSING. Here the issue starts because GROUP BY is a time consuming
+operation because the data is not already correctly pre-sorted by the scan.
```
Change that to

GROUP BY is a time consuming
operation. This is made even worse because ...

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
````diff
@@ -0,0 +1,201 @@
+ENFORCE LOCAL INDEX ON  TEST.T1 (sid);
+```
+
+Let's run the following statement to see if the zonemap was used.
````
zonemap -> zone map

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
````diff
@@ -0,0 +1,201 @@
+Let's run the following statement to see if the zonemap was used.
+
+```sql
+SELECT DISTINCT sid
````
I would recommend to use GROUP BY instead of distinct in the examples.

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+
+## Explanation
+
+The engine computes the data based on the execution plan created by the compiler. During filter stages, the Engine decides on the use of zone maps.
```
This needs a more extensive explanation.

The compiler sees the existence of the index on SID. Due to that it can create a SCAN that optimizes the pipeline for the GROUP BY. Due to that, the data layout changes and the Engine decides not to use zone maps for the filtering.

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+The SCAN uses the index on SID to identify the best ROWID order for the GROUP BY. So the data block layout after
+the SCAN is:
+
+|ROWID|TIMES_SEGMENT_ID|TIMES_VALUE|SID_VALUE|
```
TIMES_VALUE -> TIME_ZONES
SID_VALUE -> SID

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+Now let us continue with the example.
+After the PIPE FILTER filter stage down the data block layout is:
+
+|ROWID|TIMES_SEGMENT_ID|TIMES_VALUE|SID|
```
TIMES_VALUE -> TIME_ZONES

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+|128|5|'2025-06-04'|20717|
+
+The input is ordered as above. This means, engine reads the first entries and sees TIMES_SEGMENT_ID 1, 6, 5, 1, ...
+Accessing the metadata of a segment has an overhead and also consumes memory.  So our engine only accesses the metadata if consecutive rows in the data block have the same SEGMENT_ID.
```
Accessing the metadata of a segment has an overhead and also consumes memory.  So our engine only accesses the metadata if consecutive rows in the data block have the same SEGMENT_ID.

->

Accessing the metadata of a segment has an overhead and also consumes memory.  In the above example, the data is ordered by SID. Thus, there are not many consecutive rows with the same data block. Also there are many distinct SEGMENT_IDs. Accessing and switching between zone map data has a runtime and memory overhead. So for every block of data, engine can decide whether it is more beneficial to use zone maps or not. If Engine uses zone maps on at least one block of data, the remark is added to the profiling.

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+|128|5|'2025-06-04'|20717|
+|65|6|'2025-06-05'|5|
+
+Let us assume the zone maps for the TIMES_SEGEMENT_IDs (1,2,3,4) have min = '2020-01-01' and max = '2020-01-01'.
```
TIMES_SEGEMENT_IDs -> TIMES_SEGMENT_IDs

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+
+Let us assume the zone maps for the TIMES_SEGEMENT_IDs (1,2,3,4) have min = '2020-01-01' and max = '2020-01-01'.
+This means the filter part of the scan can remove the segments 1,2,3,4 with zone maps and only needs to filter the rest.
+Thus the data block layout for the next stage after the filter part of FILTERSCAN is
```
Thus the data block layout
->
Thus for our simplified example, the data block layout

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
````diff
@@ -0,0 +1,201 @@
+ENFORCE ZONEMAP ON TEST.T1 (time_zones);
+```
+
+and add an index to the SID, which in real use cases can be created by join queries:
````
add a local index on column SID

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+
+Let's break down why this can happen in Exasol:
+
+Let us do the following test, and lets remove the index:
```
lets remove
->
let's remove

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+|128|5|'2025-06-04'|20717|
+
+> [!IMPORTANT]
+> This is the perfect input for the GROUP BY and considerable speeds up the following GROUP BY computation.
```
considerable -> considerably

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+|128|5|'2025-06-04'|20717|
+|65|6|'2025-06-05'|$${\color{red}5}$$|
+
+Now we do a GROUP BY POSTPROCESSING. Here the issue starts because GROUP BY is a time consuming
```
time consuming -> time-consuming

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
```diff
@@ -0,0 +1,201 @@
+-- data model preparation
+CREATE SCHEMA TEST;
+
+CREATE OR REPLACE TABLE TEST.T1  (time_zones timestamp, sid int);
```
I would prefer time_value over time_zones.

If we change it here, we also have to change the text below.

narmion commented on Database-Features/zone-map-usage-optimizer-question.md:
````diff
@@ -0,0 +1,201 @@
+```sql
+SELECT DISTINCT sid
+FROM TEST.T1
+WHERE time_zones between '2021-06-01 00:00:00' and '2025-06-06 00:00:00';
````
Based on the input data, this is not a very strong filter. I am not sure whether this is intentional.

narmion reviewed (commented):
I think it makes sense if a second person reads this again after changing it. I know to much about the backgrounds to be a neutral reviewer.
